Rename schema to database in PostgresHook

feluelle opened the pull request:
In PostgresHook the "schema" field is only being called like that to make it compatible with the underlying DbApiHook which uses the schema for the sql alchemy connector. The postgres connector library however does not allow setting a schema in the connection instead a database can be set. To clarify that, we change all references in the PostgresHook code and documentation.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

feluelle committed:
Rename schema to database in PostgresHook

In PostgresHook the "schema" field is only being called like that to make it compatible with the underlying DbApiHook which uses the schema for the sql alchemy connector. The postgres connector library however does not allow setting a schema in the connection instead a database can be set. To clarify that, we change all references in the PostgresHook code and documentation.

feluelle commented on docs/apache-airflow-providers-postgres/connections/postgres.rst:
```diff
@@ -29,7 +29,7 @@ Host (required)
 
 Schema (optional)
-    Specify the schema name to be used in the database.
+    Specify the database name.
```
Note that the field in the connection will still be `schema`. This is due to how we build the connection string for all databases. But I updated the docs to make sure that it should contain the database name.

kaxil replied:
Should we add a reason in the docs as well on why we are using `schema` field to set `database`?

feluelle replied:
Thanks @kaxil I have updated the docs. PTAL :)

jedcunningham committed:
Merge branch 'main' into fix/postgres-schema-database-inconsistency

kaxil commented on airflow/providers/postgres/hooks/postgres.py:
```diff
@@ -95,7 +100,7 @@ def get_conn(self) -> connection:
             user=conn.login,
             password=conn.password,
-            dbname=self.schema or conn.schema,
+            dbname=self.database or conn.schema,
```
Do we still want to set `dbanem` from `conn.schema`?

kaxil replied:
Is this mainly for backwards-compat?

kaxil replied:
The following should work for setting schema?
```
postgresql://localhost/airflow?options=-csearch_path%3Ddbo,public
```

feluelle replied:
Interesting 🤔 I will test that.

feluelle replied:
I don't want to add a new feature for adding this in the connection. We can add that separately if really necesseary. 

I have documented how users can set the default schema as an additional query or use PostgresOperator runtime params to run the searchpath option before the queries get executed.

potiuk replied:
Agree. I think it's a simple/clever way of breaking of the schema/database naming issues :)

feluelle committed:
Make change backwards-compatible

- use deprecation warning instead of log.error
- clarify docs about database & schema

feluelle reviewed (commented):
I added backwards compatibility. Please review again @kaxil. 🙏

feluelle commented on airflow/providers/postgres/operators/postgres.py:
```diff
@@ -42,6 +42,8 @@ class PostgresOperator(BaseOperator):
     :param parameters: (optional) the parameters to render the SQL query with.
     :param database: name of database which overwrite defined one in connection
+    :param runtime_parameters: a mapping of runtime params added to the final sql being executed.
+        E.g you could set the schema via {"search_path": "CUSTOM_SCHEMA"}
```
```suggestion
        For example, you could set the schema via `{"search_path": "CUSTOM_SCHEMA"}`.
```

feluelle committed:
Fix doc issue

feluelle committed:
Update airflow/providers/postgres/operators/postgres.py

feluelle committed:
Merge branch 'main' into fix/postgres-schema-database-inconsistency

feluelle commented on airflow/providers/postgres/hooks/postgres.py:
```diff
@@ -67,10 +68,17 @@ class PostgresHook(DbApiHook):
+            warnings.warn(
+                'The "schema" arg has been renamed to "database" as it contained the database name.'
+                'Please use "database" to set the database name.',
+                DeprecationWarning,
```
```suggestion
                DeprecationWarning,
                stacklevel=2,
```

feluelle committed:
Update airflow/providers/postgres/hooks/postgres.py

potiuk commented on docs/apache-airflow-providers-postgres/connections/postgres.rst:
```diff
@@ -29,7 +29,7 @@ Host (required)
 
 Schema (optional)
-    Specify the schema name to be used in the database.
+    Specify the name of the database to connect to. Note that a database schema is not allowed during connect. Instead you can use ``SET search_path TO custom_schema`` if you want to change the default schema being used in a session for your queries. If you are using ``PostgresOperator`` see :ref:`Passing Server Configuration Parameters into PostgresOperator <howto/operators:postgres>`.
```
One comment here. I think it needs a bit more explanation here. What does it mean for the user that "database schema is not allowed during connect"  - I think it should be described a bit more from the user perspective: "If you do not use Postgres operator but run query directly via Postgres Hook, you need also to set  schema via "SET search_path". ... Or comething like that (if I understand correctly the consequences)

feluelle replied:
I have updated the docs in 8e9b1ba to reference postgres docs. I think this makes it clearer & cleaner. WDYT?

potiuk replied:
Looks better :)

feluelle committed:
Update docs

potiuk reviewed (approved):
LGTM. WDYT @kaxil ?

ashb commented on airflow/providers/postgres/hooks/postgres.py:
```diff
@@ -67,10 +68,18 @@ class PostgresHook(DbApiHook):
+                DeprecationWarning,
+                stacklevel=2,
+            )
+            kwargs['database'] = kwargs['schema']
```
Please can you add a test that ensures this still works

feluelle replied:
Done. PTAL :)

feluelle committed:
Add test for schema database kwarg compatibility
